include/drivers/sensor.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#ifndef DRIVERS_SENSOR_H
+#define DRIVERS_SENSOR_H
+
+#include <dm.h>
+#include <intrusive.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <util.h>
+
+#define SENSOR_OPS(dev) \
+	(&container_of((dev)->drv, struct sensor_driver, drv)->ops)
+
+struct sensor_info {
+	const char *const name;    /**< Name exported to SCPI. */
+	uint32_t          reading; /**< Current value from sensor. */
+	const uint8_t     address; /**< address of register. */
+};
+
+struct sensor_driver_ops {
+	uint8_t             (*get_count)(struct device *dev);
+	struct sensor_info *(*get_info)(struct device *dev, uint8_t id);
+	int                 (*read_raw)(struct device *dev, uint8_t id,
+	                                uint32_t *raw);
+};
+
+struct sensor_driver {
+	const struct driver            drv;
+	const struct sensor_driver_ops ops;
+};
+
+/**
+ * Get the number of sensors within a device.
+ *
+ * This function has no defined errors.
+ *
+ * @param dev   A device containing one or more sensors.
+ * @return      The number of sensors available on this device.
+ */
+static inline uint8_t
+sensor_get_count(struct device *dev)
+{
+	return SENSOR_OPS(dev)->get_count(dev);
+}
+
+/**
+ * Get generic information about a sensor.
+ *
+ * This function has no defined errors.
+ *
+ * @param dev   The device containing this sensor.
+ * @param id    The device-specific identifier for this sensor.
+ * @return      A pointer to the information structure.
+ */
+static inline struct sensor_info *
+sensor_get_info(struct device *dev, uint8_t id)
+{
+	return SENSOR_OPS(dev)->get_info(dev, id);
+}
+
+/**
+ * Get the current value of a sensor. If the sensor is disabled, this
+ * function returns the value the sensor would have if it was enabled. If
+ * the value is unknown, this function returns zero for the value
+ *
+ * This function may fail with:
+ *   EIO    There was a problem communicating with the hardware.
+ *
+ * @param dev   The device containing this sensor.
+ * @param id    The device-specific identifier for this sensor.
+ * @param value The location to store the value read from the sensor.
+ * @return      Zero on success; a defined error code on failure.
+ */
+int sensor_get_value(struct device *dev, uint8_t id, uint16_t *value);
```

Comment:
This needs to be implemented.

Reply:
what is the difference between sensor_get_value and read_raw?

Reply:
One returns the scaled ("cooked") value in standard units (millivolts, millicelsius, microamps, etc.). The `read_raw` function simply returns the contents of the register (only the bits that represent a value). That way we only implement scaling once.

Reply:
done

Reply:
done

Reply:
Likewise, `value` needs to be a `uint32_t *` here.

Reply:
done